sklearn/metrics/_pairwise_distances_reduction/_radius_neighbors.pyx.tp

```diff
@@ -84,8 +84,7 @@ cdef class RadiusNeighbors{{name_suffix}}(BaseDistancesReduction{{name_suffix}})
         """
         if (
             metric in ("euclidean", "sqeuclidean")
-            and not issparse(X)
-            and not issparse(Y)
+            and not (issparse(X) ^ issparse(Y))
```

Comment:
Side-note: I think it would be nice to turn in another PR `is_valid_sparse_matrix`, i.e.:

https://github.com/scikit-learn/scikit-learn/blob/40a6afb4a5f1c0da43723653fb24bf2a4f6a3f77/sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py#L101-L111

into `is_supported_sparse_matrix` in the global scope of `base.pyx` and propagated in place of `scipy.sparse.{issparse,isspmatrix_csr}` in `sklearn.metrics._pairwise_distances_reduction`.